FEXCore/Source/Interface/Core/CodeCache.cpp

```diff
@@ -329,6 +323,156 @@ bool CodeCache::SaveData(Core::InternalThreadState& Thread, int fd, const Execut
+    Allocator::VirtualDontNeed(Thread.CallRetStackBase, FEXCore::Core::InternalThreadState::CALLRET_STACK_SIZE);
+    auto lk = Thread.LookupCache->AcquireWriteLock();
+    Thread.LookupCache->ChangeGuestToHostMapping(*Prev, *CTX.GetLatest()->LookupCache, lk);
+  }
+
+  auto CodeBuffer = CTX.GetLatest();
+  LOGMAN_THROW_A_FMT(header.CodeBufferSize <= CodeBuffer->Size, "CodeBuffer too small to load code cache");
+  LOGMAN_THROW_A_FMT(reinterpret_cast<uintptr_t>(CodeBuffer->Ptr) % 0x1000 == 0, "Expected CodeBuffer base to be page-aligned");
+  const auto Delta = AlignUp(CTX.LatestOffset, 0x1000) - CTX.LatestOffset;
+  CTX.LatestOffset += Delta;
+
+  while (CTX.LatestOffset + header.CodeBufferSize > CodeBuffer->Size - Utils::FEX_PAGE_SIZE) {
+    CTX.ClearCodeCache(&Thread);
+    CodeBuffer = CTX.GetLatest();
+    LogMan::Msg::IFmt("Increased code buffer size to {} MiB for cache load", CodeBuffer->Size / 1024 / 1024);
+  }
+
+  // Read CodeBuffer data from file. Make sure the destination is page-aligned.
+  // TODO: Only load the data needed for the selected section
+  auto CodeBufferRange = std::as_writable_bytes(std::span {CodeBuffer->Ptr, CodeBuffer->Size}).subspan(CTX.LatestOffset, header.CodeBufferSize);
+  ::memcpy(CodeBufferRange.data(), MappedCacheFile, header.CodeBufferSize);
+  MappedCacheFile += header.CodeBufferSize;
+  CTX.LatestOffset += header.CodeBufferSize;
+
+  // Apply FEX relocations
+  auto Ret = ApplyCodeRelocations(BinarySection.FileStartVA, CodeBufferRange, Relocations, false);
+  LOGMAN_THROW_A_FMT(Ret == true, "Failed to apply code cache relocations");
+
+  // Register blocks to LookupCache
+  {
+    auto& LookupCache = *CodeBuffer->LookupCache;
+    for (auto& [Guest, Host] : BlockList) {
+      LookupCache.BlockList[Guest + BinarySection.FileStartVA].HostCode = Host.HostCode + reinterpret_cast<uintptr_t>(CodeBufferRange.data());
+
+      for (auto& CodePage : Host.CodePages) {
+        CodePage += BinarySection.FileStartVA;
+      }
+      LookupCache.BlockList[Guest + BinarySection.FileStartVA].CodePages = std::move(Host.CodePages);
+    }
+  }
+
+  // Register loaded code ranges
+  fextl::vector<uint64_t> Entrypoints;
+  fextl::set<uint64_t> EntrypointsSet;
+  for (uint32_t i = 0; i < header.NumCodePages; ++i) {
+    uint64_t CodePage;
+    memcpy(&CodePage, MappedCacheFile, sizeof(CodePage));
+    MappedCacheFile += sizeof(CodePage);
+
+    uint64_t NumEntrypoints;
+    memcpy(&NumEntrypoints, MappedCacheFile, sizeof(NumEntrypoints));
+    MappedCacheFile += sizeof(NumEntrypoints);
+
+    Entrypoints.resize(NumEntrypoints);
+    memcpy(Entrypoints.data(), MappedCacheFile, NumEntrypoints * sizeof(Entrypoints[0]));
+    MappedCacheFile += NumEntrypoints * sizeof(Entrypoints[0]);
+
+    EntrypointsSet.clear();
+    EntrypointsSet.insert(Entrypoints.begin(), Entrypoints.end());
+    if (Thread.LookupCache->AddBlockExecutableRange(&Thread, EntrypointsSet, CodePage, FEXCore::Utils::FEX_PAGE_SIZE)) {
```

Comment:
Can switch this to use iterators to avoid the intermediary set

Reply:
It's already outside :)

Will move the `fextl::set` back in though, since I don't think that container can benefit from reusing allocations.

Reply:
> I think I'd prefer deferring this since type-erasing over iterators is annoying to implement, and I'm not sure it makes a significant dent in performance given we're in IO code.

I was looking at this and a little confused, can one literally not just template both funcs?

Reply:
Also - this should use CodeBuffer LookupCache->AddBlockExecutableRange rather than going through thread